### PR TITLE
Add new k8s apiserver event source

### PR DESCRIPTION
**Pull request by lionelvillard**
## Proposed Changes

  * Add k8s object event source
  * Add ApiServerSource CRD for configuring k8s apiserver resource watchers

### CODE REVIEW

**Comment by lionelvillard** [comment]
@grantr @Harwayne this PR is now ready for review.

**Comment by aslom** [comment]
I tested PR and it worked. Instructions for kail may need improvements ended up using kubectl logs as kail was not showing any changes.

File: contrib/etcd/cmd/controller/main.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
+)
+
+func main() {
+	var metricsAddr string
+	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
```
Comment (n3wscott) [inline contrib/etcd/cmd/controller/main.go]: knative has been using 9090 for metrics and 8080 for cluster data plane.

File: contrib/etcd/cmd/receiver_adapter/main.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
+)
+
+func main() {
+	var metricsAddr string
+	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
```
Comment (n3wscott) [inline contrib/etcd/cmd/receiver_adapter/main.go]: see above note about 8080 vs 9090

File: contrib/etcd/cmd/receiver_adapter/main.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	if err != nil {
+		log.Error(err, "unable to set up overall controller manager")
+		os.Exit(1)
+	}
+
+	log.Info("Registering Components.")
+
+	// Setup all Controllers
+	log.Info("Setting up controller")
+	if err := adapter.Add(mgr); err != nil {
```
Comment (n3wscott) [inline contrib/etcd/cmd/receiver_adapter/main.go]: Can you detail why the receive adaptor is being run like a controller? How are the controller and receive adaptors going to be used in this source?

File: contrib/etcd/pkg/apis/sources/v1alpha1/etcdsource_types.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+)
+
+var etcdCondSet = duckv1alpha1.NewLivingConditionSet(
+	EtcdConditionSinkProvided,
+	EtcdConditionDeployed,
+)
+
+// EtcdSourceSpec defines the desired state of EtcdSource
+type EtcdSourceSpec struct {
+	// Resources is the list of resources to watch
```
Comment (n3wscott) [inline contrib/etcd/pkg/apis/sources/v1alpha1/etcdsource_types.go]: ooohhhh ok, I think this source is misnamed. It is not etcd you are sourcing, it is the k8s api server.

File: contrib/etcd/samples/clusterrolebinding.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+metadata:
+  name: etcd-event-source-sample
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: etcd-event-source-sample
+subjects:
+- kind: ServiceAccount
+  name: etcd-event-source-sample
+  namespace: default
```
Comment (n3wscott) [inline contrib/etcd/samples/clusterrolebinding.yaml]: nit: newline

File: contrib/etcd/samples/etcdsource.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  name: etcdsource-sample
+spec:
+  serviceAccountName: etcd-event-source-sample
+  resources:
+    - apiVersion: v1
+      kind: ConfigMap
+  sink:
+    apiVersion: serving.knative.dev/v1alpha1
+    kind: Service
+    name: event-display
```
Comment (n3wscott) [inline contrib/etcd/samples/etcdsource.yaml]: nit: newline

File: contrib/etcd/pkg/adapter/adapter.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	eventsClient, err := kncloudevents.NewDefaultClient(sinkURI)
+	if err != nil {
+		return err
+	}
+
+	// Create one controller per watch
+	i := 0
+	for {
```
Comment (n3wscott) [inline contrib/etcd/pkg/adapter/adapter.go]: This is clever and maybe you would like to contribute it to https://github.com/kelseyhightower/envconfig

File: contrib/etcd/pkg/adapter/adapter.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+		logger.Errorf("could not find %s %v\n", request.Name, request)
+		return reconcile.Result{}, nil
+	}
+	eventType := updateEventType
+	if object.GetDeletionTimestamp() != nil {
+		eventType = deleteEventType
+	}
+
+	event := cloudevents.Event{
+		Context: cloudevents.EventContextV03{
```
Comment (n3wscott) [inline contrib/etcd/pkg/adapter/adapter.go]: V03 is not released yet (the cloudevents spec is still changing). Please use V02 for now.

File: contrib/etcd/pkg/adapter/adapter.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+			Type:   eventType,
+			Source: *types.ParseURLRef("/Etcd"),
```
Comment (n3wscott) [inline contrib/etcd/pkg/adapter/adapter.go]: the source should at the very least the namespace, name, gvk of the resource in question.

**Comment by lionelvillard** [comment]
@n3wscott thanks for your review

> I think it is mis-named, I was assuming it would interact with etcd, ala serverless.coms event gateway. But it really does resource watches on the k8s api server.

Totally agree.  

> So perhaps this can be folded into the k8s event source? Then that source can sign up for k8s events emitted by a resource, and resource mutations as you have implemented. What do you think @vaikas-google ?

Indeed, this event provider subsumes the k8s event source.

File: contrib/etcd/cmd/receiver_adapter/main.go (hunk 3 above)
Comment (lionelvillard) [reply contrib/etcd/cmd/receiver_adapter/main.go]: In this case the receive adaptor is just like a controller: it watches for k8s events and the reconcile method is about sending the cloud event. The controller runtime takes care of the machinery and the receive adaptor only worries about `reconcile`.

This source works the same as as the other sources. It's mostly copy&paste of `gcpubsub`.

File: contrib/etcd/pkg/adapter/adapter.go (hunk 9 above)
Comment (lionelvillard) [reply contrib/etcd/pkg/adapter/adapter.go]: what about doing the same as what `kubernetesevents` is doing:

```
ID:     string(m.ObjectMeta.UID),
Source: *types.ParseURLRef(createSelfLink(m.InvolvedObject)),
Time:   &types.Timestamp{Time: m.ObjectMeta.CreationTimestamp.Time},
```

**Comment by lionelvillard** [comment]
@n3wscott I changed the name to ApiServer. Hopefully this works for you.

What else needs to be done for this PR to get approved and merged?

**Comment by lionelvillard** [comment]
@n3wscott the tests seem to be stuck somewhere (I submitted an update 2 hours ago). Can you take a look?

**Pull request merged**